tf2/src/buffer_core.cpp

```diff
@@ -114,49 +114,62 @@ std::string stripSlash(const std::string& in)
   return out;
 }
 
-
-bool BufferCore::warnFrameId(const char* function_name_arg, const std::string& frame_id) const
+CompactFrameID BufferCore::validateFrameId(
+  const char* function_name_arg,
+  const std::string& frame_id,
+  std::string* error_msg) const
 {
-  if (frame_id.size() == 0)
+  std::stringstream ss;
```

Comment:
> This was found to be a large bottleneck in past performance testing.

In this case, does there even need to be a string stream? I think this would work too, since we have C++14.

```cpp
using namespace std::string_literals;

if (nullptr != error_msg) {
  *error_msg = "Invalid frame ID \""s + frame_id + "\" passed to "s + function_name_arg + " - "s + rationale;
}
```

Reply:
@allenh1 to maintain existing functionality I need to construct the message for either the `*error_msg =` case or the `CONSOLE_BRIDGE_logWarn` case - we shouldn't duplicate that code, so I think I need to assign it to a variable of some sort anyways. Instead of speculating about performance I constructed a quick test of various message construction techniques, assuming we have to make a local string for warning, rather than just filling *error_msg.

```
#include <iostream>
#include <string>
#include <chrono>
#include <sstream>

#define ITERATIONS 1000000

void testStrategy(const char* name, auto callable) {
    typedef std::chrono::high_resolution_clock clock;
    typedef std::chrono::duration<float, std::milli> mil;
    auto t0 = clock::now();
    std::cout << "Testing strategy \"" << name << "\":" << std::endl;
    const char * function_name_arg = "lookupTransform argument target_frame";
    const std::string frame_id = "base_link";
    const char * rationale = "frame does not exist";

    std::string out;
    for (int i = 0; i < ITERATIONS; i++) {
        out = callable(function_name_arg, frame_id, rationale);
    }

    auto t1 = clock::now();
    std::cout << "  " << mil(t1-t0).count() << "ms" << std::endl;
}


int main ()
{

    testStrategy("singleLinePlus",
        [](
            const char * function_name_arg,
            const std::string frame_id,
            const char* rationale
        ) {
            return "Invalid frame Id \"" + frame_id + "\" passed to " + function_name_arg + " - " + rationale;
        });

    testStrategy("multiLinePlus",
        [](
            const char * function_name_arg,
            const std::string frame_id,
            const char* rationale
        ) {
            std::string tempout = "Invalid frame id \"";
            tempout += frame_id;
            tempout += "\" passed to ";
            tempout += function_name_arg;
            tempout += " - ";
            tempout += rationale;
            return tempout;
        });

    testStrategy("append",
        [](
            const char * function_name_arg,
            const std::string frame_id,
            const char* rationale
        ) {
            std::string tempout = "Invalid frame id \"";
            tempout.append(frame_id)
            .append("\" passed to ")
            .append(function_name_arg)
            .append(" - ")
            .append(rationale);
            return tempout;
        });

    testStrategy("stringstream",
        [](
            const char * function_name_arg,
            const std::string frame_id,
            const char* rationale
        ) {
            std::stringstream ss;
            ss << "Invalid frame Id \"" << frame_id << "\" passed to " << function_name_arg << " - " << rationale;
            return ss.str();
        });
}
```


```
# g++ -O0
Testing strategy "stringstream":
  412.258ms
Testing strategy "append":
  113.275ms
Testing strategy "multiLinePlus":
  111.415ms
Testing strategy "singleLinePlus":
  148.815ms
```

```
# g++ -O1
Testing strategy "stringstream":
  420.088ms
Testing strategy "append":
  85.7351ms
Testing strategy "multiLinePlus":
  88.1533ms
Testing strategy "singleLinePlus":
  80.898ms
```

```
# g++ -O2
Testing strategy "stringstream":
  390.024ms
Testing strategy "append":
  88.1943ms
Testing strategy "multiLinePlus":
  87.2995ms
Testing strategy "singleLinePlus":
  79.8148ms
```


It is clear that creating a stringstream is the slowest approach by far. Strangely the single-line addition is slower than the other options in non-optimized builds, but in optimized builds it's consistently just a little bit faster in the context of this test. I'm no compiler whiz, so it's possible that the optimizer is taking advantage of the simplicity of the structure of this test. 

I think it's safe to assume that any of those 3 methods are practically equal for our use case. I'll update to remove the use of the stringstream here.